raft/log.hh

```diff
@@ -24,6 +24,8 @@ class log {
     snapshot_descriptor _snapshot;
     // We need something that can be truncated from both sides.
     // std::deque move constructor is not nothrow hence cannot be used
+    // also, boost::deque deallocates blocks when items are removed,
+    // we don't want to hold on to memory we don't use.
```

Comment:
I am still strongly against mentioning this in the comments only. It is very easy to make sure this holds true no matter what `boost::deque` does internally. Just call fit_to_shrink(). If `log_entries` changes to a container that does not support it a *compiler* will tell you that you need to change the code accordingly. 

BTW here you do not even see that `log_entries` is `boost::deque` we try to abstract it. whoever will change `log_entries` does not have to see this comment at all, so the comment should be above `log_entries` definition

It is also not reasonable to rely on internals of a library we do not bundle. If other version has different behaviour a person who compiles against it will not get the behaviour you mention in the comment. Compiler are notoriously bad in reading comments.

Reply:
> BTW here you do not even see that log_entries is boost::deque we try to abstract it. whoever will change log_entries does not have to see this comment at all, so the comment should be above log_entries definition

That's why the comment is duplicated here and above the definition.

Generally, I don't agree with you. My points:
* it's idealistic that someone can just switch the container implementation without a deep understanding of how it's used and what are the consequences. And if someone did the research, he can't help but see this comment.

* the tradeoffs of the boost deque are relied on by the numerous usages of this library, they can't just change the behaviour as this would break a lot of things.

* boost::deque [doesn't mention](https://www.boost.org/doc/libs/1_80_0/doc/html/boost/container/deque.html) they have any relation to std::deque

On the other hand, it doesn't seem all that important, I can add `shrink_to_fit`, but I don't see how it makes things better. Usually `shink_to_fit` is used when there is a memory shortage and a guarantee that it will not be called too often. Otherwise we can get a quadratic behaviour if the implementation has to allocate/copy on each push_back.

Reply:
> And I know what tradeoffs are from the interface description. Not by looking at the implementation.

It's oversimplification. For example, the doc for `shrink_to_fit` says:

> Effects: Tries to deallocate the excess of memory created with previous allocations. The size of the deque is unchanged
> 
> Throws: If memory allocation throws.
> 
> Complexity: Constant.

It doesn't prevent it from freeing up all the allocated and not used memory, and if does, we get quadratic complexity if we call `shrink_to_fit` on each `emplace_back`.

> Why would they?

Because it specifies an interface? Nothing else comes to my mind that we could call a `deque's` interface and consider different implementations of. `boost deque` implements only `boost deque`'s interface.

> But it makes sure that if you change the container the behaviour will either stay the same or you will have to consider what to do with freed entries. A compiler will remind you.

A compiler won't remind me of quadratic complexity.  `shrink_to_fit` is not much different from the comment here - if I replace `boost::deque` with another `deque`, which frees whatever it can on `shrink_to_fit`, the only chance for me not to get quadratic behavior is to scrutinise what the word `shrink_to_fit` really means by reading the documentation (or source code). The comment is better because it clearly articulates the tradeoffs and is located in a prominent place.

Reply:
> > And I know what tradeoffs are from the interface description. Not by looking at the implementation.
> 
> It's oversimplification. For example, the doc for `shrink_to_fit` says:

I do not see what is oversimplified about it. We knew the data structure does not satisfy all our requirements, but it was close enough and there was no point starting to write raft by writing a low level container. Deque was good enough to bootstrap the project. 

> 
> > Effects: Tries to deallocate the excess of memory created with previous allocations. The size of the deque is unchanged
> > Throws: If memory allocation throws.
> > Complexity: Constant.
> 
> It doesn't prevent it from freeing up all the allocated and not used memory, and if does, we get quadratic complexity if we call `shrink_to_fit` on each `emplace_back`.

Why would we call `shrink_to_fit` on each `emplace_back`? Why would it cause quadratic complexity if we would?

> 
> > Why would they?
> 
> Because it specifies an interface? Nothing else comes to my mind that we could call a `deque's` interface and consider different implementations of. `boost deque` implements only `boost deque`'s interface.

The name deque specifies the interface. Should it list all implementations? And I have to say it is a surrealistic feeling when I was the one who choose boost::deque and definitely considered different implementations (there is even a comment about it) and now here you try to prove to me that this is not what happened :)

> 
> > But it makes sure that if you change the container the behaviour will either stay the same or you will have to consider what to do with freed entries. A compiler will remind you.
> 
> A compiler won't remind me of quadratic complexity.  `shrink_to_fit` is not much different from the comment here - if I replace `boost::deque` with another `deque`, which frees whatever it can on `shrink_to_fit`, the only chance for me not to get quadratic behavior is to scrutinise what the word `shrink_to_fit` really means by reading the documentation (or source code). The comment is better because it clearly articulates the tradeoffs and is located in a prominent place.

Reading documentation is a good idea anyway. Are you going to replace a container without reading it? Is `shrink_to_fit`  is your only concern and you do not care about the complexity of adding and removing elements? Why do you single out `shrink_to_fit`  then? What having `shrink_to_fit` in the code will tell you is that is the desired behaviour and you need to consider its complexity.

Reply:
I think it's safe to assume that the memory usage of `boost::deque` is linear w.r.t number of elements (assuming we don't manually reserve more capacity). Nothing in the documentation states that explicitly, but meh.

>  It is very easy to make sure this holds true no matter what boost::deque does internally. Just call fit_to_shrink(). 

`shrink_to_fit` ensures nothing:
```
Effects: Tries to deallocate the excess of memory created with previous allocations. The size of the deque is unchanged
```
it "tries to" deallocate, so it's not really a guarantee.

Anyway, if the comment is a problem, can we just remove it and move on?

Reply:
> I think it's safe to assume that the memory usage of `boost::deque` is linear w.r.t number of elements (assuming we don't manually reserve more capacity). Nothing in the documentation states that explicitly, but meh.

It is safe to assume that it is liner w.r.t max number of elements that was ever inserted into it. But this is exactly what we do not want to happen. We want it to be liner w.r.t current number of elements. `boost::deque` is in its current implementation. I do not want to rely on its internals. That was a random choice anyway.

> 
> > It is very easy to make sure this holds true no matter what boost::deque does internally. Just call fit_to_shrink().
> 
> `shrink_to_fit` ensures nothing:
> 
> ```
> Effects: Tries to deallocate the excess of memory created with previous allocations. The size of the deque is unchanged
> ```
> 
> it "tries to" deallocate, so it's not really a guarantee.

It tries to deallocate memory. That what we want on log truncate. If it can't it can't. If it will prove to be a problem for us we will replace it with something that do it better. `boost::deque` was chosen as a compromise, not because if fitted our  requirements perfectly.

> 
> Anyway, if the comment is a problem, can we just remove it and move on?

The comment is not the problem. I want to reflect in the code what is written in the comment.

Reply:
> Why would we call shrink_to_fit on each emplace_back? Why would it cause quadratic complexity if we would?

If we call `shrink_to_fit` for each `emplace_back` the `deque` will potentially need to allocate/copy the entire data structure just to have room to place the next item, when the next `emplace_back` comes.

I feel like we are going in circles with this discussion. Where exactly, according to your suggestion, should I call `shrink_to_fit'?

Reply:
> > Why would we call shrink_to_fit on each emplace_back? Why would it cause quadratic complexity if we would?
> 
> If we call `shrink_to_fit` for each `emplace_back` the `deque` will potentially need to allocate/copy the entire data structure just to have room to place the next item, when the next `emplace_back` comes.

First of all we want to call it only on log truncation. not on `emplace_back` . Second for the deque the complexity of the `emplace_back` at the beginning and at the end is well defined. It is O(1). This is one of the reasons why it was chosen. 

> 
> I feel like we are going in circles with this discussion. Where exactly, according to your suggestion, should I call `shrink_to_fit`?

In log truncation only.

Reply:
> Second for the deque the complexity of the emplace_back at the beginning and at the end is well defined. It is O(1).

It's not `O(1)`, it's amortized O(1):
> template<class... Args> reference emplace_back(Args &&... args);
> Effects: Inserts an object of type T constructed with std::forward<Args>(args)... in the end of the deque.
> 
> Returns: A reference to the created object.
> 
> Throws: If memory allocation throws or the in-place constructor throws.
> 
> Complexity: Amortized constant time

And amortization only works if we reuse allocated memory.

Ok, I will add it after each `_log.erase`. There are two of them, `apply_snapshot` and `truncate_uncommitted`.

Reply:
> > Second for the deque the complexity of the emplace_back at the beginning and at the end is well defined. It is O(1).
> 
> It's not `O(1)`, it's amortized O(1):
https://en.cppreference.com/w/cpp/container/deque says:
Insertion or removal of elements at the end or beginning - constant O(1)

Yes, it is amortized, but by such large factor that cppreference don't even consider it worth mentioning.

> 
> > template<class... Args> reference emplace_back(Args &&... args);
> > Effects: Inserts an object of type T constructed with std::forward(args)... in the end of the deque.
> > Returns: A reference to the created object.
> > Throws: If memory allocation throws or the in-place constructor throws.
> > Complexity: Amortized constant time
> 
> And amortization only works if we reuse allocated memory.
> 
> Ok, I will add it after each `_log.erase`. There are two of them, `apply_snapshot` and `truncate_uncommitted`.

`apply_snapshot` is the important one.